### DISCUSSION TITLE
Cycle count directive

### DISCUSSION THREAD

**Discussion by Pokeman2003** [vhelin/wla-dx#647] [Ideas]
This is an idea probably not even good on paper and is very likely totally hell to practically implement. It also is unlikely to be particularly useful outside of the VERY specific use case of Atari 2600 programming, where counting cycles is incredibly important.

In short, the directive would allow you to define a start/end label, and then return how many clock cycles that occur between the two. The default behavior, if the labels were not defined, would be the first encountered label backwards/start of segment, and the directive itself. The directive would obviously error if any of the code could be a variable number of cycles(such as a MUL.) It would also assume that any conditional branching is read but not executed. 
Probably the easiest way to implement this on the instruction side is to just assign a integer value for each instruction which resembles the cycle count, use that to create the count. You can throw an error if any of the instructions are 0 cycles long, and then assign multi-length directives that number.

**Comment by vhelin** [comment]
Interesting idea, maybe with an option to print out assembly source lines where there'd be per line the cycle count as a comment?

Implementing this for all supported CPUs is quite a big task, but I guess we could start from popular CPUs like 6502?

**Reply by Pokeman2003** [reply]
Actually that printout might not be a bad idea. To be honest, I don't even know how you'd go about implementing a directive like this, I just figured I'd throw out my half-baked idea and see if it could get a bit more cooking time from someone more experienced.

I already recognized implementing this would be a major hassle, and I half expected there to be no one trying to take on this task. I can't even tell you how many cycles are in whatever nop-like instruction the SuperFX chip has. Thankfully some of these chips are really, REALLY well documented. The 6502 not only doesn't have repeating instructions, there are like 80 different lists of how many cycles each instruction takes, so it's a very good starting point.
